File: gto-support-core/src/main/java/org/ccci/gto/android/common/util/BundleUtils.java
package org.ccci.gto.android.common.util;

import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.ccci.gto.android.common.compat.util.LocaleCompat;

import java.lang.reflect.Array;
import java.util.Locale;

public class BundleUtils {
    @Nullable
    @SuppressWarnings({"unchecked", "SuspiciousSystemArraycopy"})
    public static <T extends Parcelable> T[] getParcelableArray(@NonNull final Bundle bundle,
                                                                @Nullable final String key,
                                                                @NonNull final Class<T> clazz) {
        final Parcelable[] raw = bundle.getParcelableArray(key);
        if (raw == null) {
            return null;
        }

        // copy all objects to typed array
        final T[] arr = (T[]) Array.newInstance(clazz, raw.length);
        System.arraycopy(raw, 0, arr, 0, raw.length);
        return arr;
    }

    public static void putEnum(@NonNull final Bundle bundle, @Nullable final String key,
                               @Nullable final Enum<?> value) {
        bundle.putString(key, value != null ? value.name() : null);
    }

    @Nullable
    public static <T extends Enum<T>> T getEnum(@NonNull final Bundle bundle, @NonNull final Class<T> type,
                                                @Nullable final String key) {
        return getEnum(bundle, type, key, null);
    }

    @Nullable
    public static <T extends Enum<T>> T getEnum(@NonNull final Bundle bundle, @NonNull final Class<T> type,
                                                @Nullable final String key, @Nullable final T defValue) {
        final String raw = bundle.getString(key);
        if (raw == null) {
            return defValue;
        }

        try {
            return Enum.valueOf(type, raw);
        } catch (final IllegalArgumentException e) {
            return defValue;
        }
    }

    public static void putLocale(@NonNull final Bundle bundle, @Nullable final String key,
                                 @Nullable final Locale locale) {
        bundle.putString(key, locale != null ? LocaleCompat.toLanguageTag(locale) : null);
    }

    @Nullable
    public static Locale getLocale(@NonNull final Bundle bundle, @Nullable final String key) {
        return getLocale(bundle, key, null);
    }

    @Nullable
    public static Locale getLocale(@NonNull final Bundle bundle, @Nullable final String key,
                                   @Nullable final Locale defValue) {
        final String raw = bundle.getString(key);
        if (raw == null) {
            return defValue;
        }
        return LocaleCompat.forLanguageTag(raw);
    }
}

File: gto-support-util/src/main/java/org/ccci/gto/android/common/util/os/BundleUtils.kt
@file:JvmName("BundleUtils")

package org.ccci.gto.android.common.util.os

import android.os.Bundle
import android.os.Parcelable
import java.util.Locale
import org.ccci.gto.android.common.compat.util.LocaleCompat
import org.jetbrains.annotations.Contract

// region Enums

fun Bundle.putEnum(key: String?, value: Enum<*>?) = putString(key, value?.name)

@JvmOverloads
@Contract("_, _, _, !null -> !null")
fun <T : Enum<T>> Bundle.getEnum(type: Class<T>, key: String?, defValue: T? = null): T? {
    return try {
        getString(key)?.let { java.lang.Enum.valueOf<T>(type, it) } ?: defValue
    } catch (e: IllegalArgumentException) {
        defValue
    }
}

@Contract("_, _, !null -> !null")
inline fun <reified T : Enum<T>> Bundle.getEnum(key: String?, defValue: T? = null) =
    getEnum(T::class.java, key, defValue)

// endregion Enums

// region Locales

fun Bundle.putLocale(key: String?, locale: Locale?) =
    putString(key, if (locale != null) LocaleCompat.toLanguageTag(locale) else null)

@JvmOverloads
@Contract("_, _, !null -> !null")
fun Bundle.getLocale(key: String?, defValue: Locale? = null) =
    getString(key)?.let { LocaleCompat.forLanguageTag(it) } ?: defValue

/**
 * Store an array of Locales in the provided Bundle
 *
 * @receiver The bundle to store the locale array in
 * @param key The key to store the locale array under
 * @param locales The locales being put in the bundle
 * @param singleString Flag indicating if the locale array should be stored as a single string
 */
@JvmOverloads
fun Bundle.putLocaleArray(key: String?, locales: Array<Locale?>?, singleString: Boolean = false) {
    val tags = locales?.map { it?.let { LocaleCompat.toLanguageTag(it) } }?.toTypedArray()

    if (singleString) {
        putString(key, tags?.joinToString(","))
    } else {
        putStringArray(key, tags)
    }
}

fun Bundle.getLocaleArray(key: String?) =
    (getStringArray(key) ?: getString(key)?.split(",")?.toTypedArray())
        ?.map { it?.let { LocaleCompat.forLanguageTag(it) } }?.toTypedArray()

// endregion Locales

// region Parcelables
@Deprecated(
    "Since v3.6.2, this was moved to the BundleKt package for java usage",
    ReplaceWith("getParcelableArray(key, clazz)")
)
@JvmName("getParcelableArray")
fun <T : Parcelable> deprecatedGetParcelableArray(bundle: Bundle, key: String?, clazz: Class<T>) =
    bundle.getParcelableArray(key, clazz)

@Deprecated(
    "Since v3.6.2, renamed to getTypedParcelableArray() for Kotlin 1.4 update",
    ReplaceWith("getTypedParcelableArray<T>(key)")
)
inline fun <reified T : Parcelable> Bundle.getParcelableArray(key: String?) = getTypedParcelableArray<T>(key)
// endregion Parcelables
